test: add v0.20.1, v0.21.0 and v22.0 to backwards compatibility test

Sjors opened the pull request:
This also simplifies the tests a bit.

Sjors committed:
test: backwards compatibility: misc fixes

This cleanup may slightly impact test coverage.

Reduce code repition by looping over the list of nodes.

Reduce brittleness by refering to nodes by name rather than index.

Add helper method nodes_wallet_dir.

MarcoFalke commented on test/functional/feature_backwards_compatibility.py:
```diff
@@ -6,7 +6,7 @@
 Test various backwards compatibility scenarios. Download the previous node binaries:
 
-contrib/devtools/previous_release.sh -b v0.19.1 v0.18.1 v0.17.1 v0.16.3 v0.15.2
+contrib/devtools/previous_release.sh -b v0.20.0 v0.19.1 v0.18.1 v0.17.1 v0.16.3 v0.15.2
```
instead of updating this in the ci config, as well as all the backward compat tests, why not simply make it the default value in the script?

MarcoFalke commented on test/functional/feature_backwards_compatibility.py:
```diff
@@ -55,13 +56,14 @@ def setup_nodes(self):
         ])
         # adapt bitcoin.conf, because older bitcoind's don't recognize config sections
-        adjust_bitcoin_conf_for_pre_17(self.nodes[5].bitcoinconf)
+        adjust_bitcoin_conf_for_pre_17(self.nodes[-1].bitcoinconf)
```
this goes away with #19294

Sjors replied:
rebased

MarcoFalke commented on test/functional/test_framework/test_framework.py:
```diff
@@ -429,7 +433,7 @@ def get_bin_from_version(version, bin_name, bin_default):
                         (version % 100) // 1,
                     ),
-                ),
+                ) + suffix,
```
not sure if we want to add dead and untested code the the test framework. This can be added when it is actually needed.

Sjors replied:
I need it every time there's a release candidate, but then it never it gets merged...

MarcoFalke replied:
Until we will be testing backwards compatibility for rcs, you can keep it as a commit in your local git and then cherry-pick as needed for your own needs.

MarcoFalke commented on test/functional/test_framework/test_node.py:
```diff
@@ -305,7 +305,11 @@ def get_wallet_rpc(self, wallet_name):
+        if self.version is None:
+            return True
+        if type(self.version) == tuple:
+            return self.version[0] >= ver
```
same

maflcko reviewed (commented):
Concept ACK

Commit:
test: Fix intermittent test failure in feature_backwards_compatibility

Sjors commented:
Rebased and dropped the release candidate commit.

MarcoFalke commented:
This fails on travis. Presumably because gpg is missing, etc

fjahr commented:
Concept ACK, code looks good, just need to fix Travis.

Also ACK on DRYing up the compatibility test.

Sjors commented:
> This fails on travis. Presumably because gpg is missing, etc

Oh, I guess we should have cleared Travis cache before merging #19205. I added a commit to fetch the release PGP key (can be cherry-picked into its own PR if we need it urgently).

MarcoFalke commented on ci/test/05_before_script.sh:
```diff
@@ -48,6 +48,7 @@ if [ -z "$NO_DEPENDS" ]; then
 fi
 if [ -n "$PREVIOUS_RELEASES_TO_DOWNLOAD" ]; then
   BEGIN_FOLD previous-versions
+  DOCKER_EXEC gpg --list-keys $MAINTAINER_KEY >/dev/null 2>&1 || gpg --keyserver hkp://keyserver.ubuntu.com:80 --recv-keys $MAINTAINER_KEY
```
```suggestion
  DOCKER_EXEC gpg --keyserver hkp://keyserver.ubuntu.com:80 --recv-keys $MAINTAINER_KEY
```

Would be surprising if a vanilla docker image had the key already

Sjors replied:
Maybe we want to cache it though if there's a DDOS on PGP key servers again.

Sjors commented:
Travis keeps failing the GPG check with a useful log entry, so instead I just added a flag to skip this check.

Sjors commented:
I'll rebase after #19813

Sjors commented:
Rebased and switched to v0.20.1

maflcko reviewed (commented):
ACK

Sjors commented on test/functional/wallet_upgradewallet.py:
```diff
@@ -6,7 +6,7 @@
 Test upgradewallet RPC. Download node binaries:
 
-test/get_previous_releases.py -b v0.19.1 v0.18.1 v0.17.2 v0.16.3 v0.15.2
+test/get_previous_releases.py -b v0.20.1 v0.19.1 v0.18.1 v0.17.2 v0.16.3 v0.15.2
```
Done, and rebased

MarcoFalke commented on test/functional/feature_backwards_compatibility.py:
```diff
@@ -4,9 +4,8 @@
-
-test/get_previous_releases.py -b v0.19.1 v0.18.1 v0.17.2 v0.16.3 v0.15.2
+Test various backwards compatibility scenarios. Requires previous releases binaries,
+see README.
```
```suggestion
see test/README.md.
```

MarcoFalke commented on test/functional/mempool_compatibility.py:
```diff
@@ -7,10 +7,7 @@
-test/get_previous_releases.py -b v0.19.1 v0.18.1 v0.17.2 v0.16.3 v0.15.2
-
-Only v0.15.2 is required by this test. The rest is used in other backwards compatibility tests.
+The previous release v0.15.2 is required by this test, see README.
```
```suggestion
The previous release v0.15.2 is required by this test, see test/README.md.
```

MarcoFalke commented on test/functional/wallet_upgradewallet.py:
```diff
@@ -6,9 +6,8 @@
-test/get_previous_releases.py -b v0.19.1 v0.18.1 v0.17.2 v0.16.3 v0.15.2
-
-Only v0.15.2 and v0.16.3 are required by this test. The others are used in feature_backwards_compatibility.py
+Requires previous releases binaries, see README.
```
```suggestion
Requires previous releases binaries, see test/README.md.
```

maflcko reviewed (commented):
ACK, but there is also a test/functional/Readme

Sjors commented:
test/README seems focussed on how to run them, whereas test/functional/README only covers development. It might make sense to move some stuff over from test/README to test/functional/README though, but maybe in another PR?

michaelfolkson commented:
ACK 0888343b6cc13d7449972345e4f576fd76a43919

Reviewed code, downloaded the previous release binaries, ran test, test passed.

Put this up on StackExchange, feel free to suggest improvements to my answer or add your own answer. https://bitcoin.stackexchange.com/questions/100924/what-backward-compatibility-testing-is-done-on-bitcoin-core

I'm also a fan of moving most the functional test docs to the functional test README and linking to it from the test README.

Will add to the possible docs improvements in https://github.com/bitcoin/bitcoin/issues/20132

Sjors committed:
test: previous releases: add v0.21.0

Sjors commented:
Rebased and simplified the code a bit. I dropped a few testing permutations in the process, but I don't think those were important.

I also added v0.21.0

hebasto commented:
Concept ACK.

hebasto reviewed (commented):
Approach ACK cb646de919fd25505f9feaa2703b1acc16cffa1a

I don't think it is required to add `*-osx.dmg`, `*-win64*`, and sources to the `test/get_previous_releases.py`.

Sjors commented:
@hebasto it doesn't hurt to commit these hashes in the repo though.

MarcoFalke commented on test/get_previous_releases.py:
```diff
@@ -59,6 +59,16 @@
+    "695fb624fa6423f5da4f443b60763dd1d77488bfe5ef63760904a7b54e91298d": "bitcoin-0.21.0-osx64.tar.gz",
+    "6223fd23d07133a6bfa2aa3d2554a09dc1d790d28ce67b0085d3fdcc1c126e05": "bitcoin-0.21.0-osx.dmg",
+    "f8b2adfeae021a672effbc7bd40d5c48d6b94e53b2dd660f787340bf1a52e4e9": "bitcoin-0.21.0-riscv64-linux-gnu.tar.gz",
+    "1a91202c62ee49fb64d57a52b8d6d01cd392fffcbef257b573800f9289655f37": "bitcoin-0.21.0.tar.gz",
```
is the source code tar.gz needed for anything? None of the other versions include it.

Sjors replied:
I'll drop it for consistency.

Sjors commented:
Slightly more complicated rebase after #20354.

Note how short the last commit is. It's now much less involved to add a new version to this test.

Sjors committed:
test: previous releases: add v22.0

Sjors commented:
Rebased and added v22.0

MarcoFalke commented on test/get_previous_releases.py:
```diff
@@ -56,9 +56,22 @@
+    "b8713c6c5f03f5258b54e9f436e2ed6d85449aa24c2c9972f91963d413e86311": "bitcoin-22.0-arm-linux-gnueabihf.tar.gz",
+    "2744d199c3343b2d94faffdfb2c94d75a630ba27301a70e47b0ad30a7e0155e9": "bitcoin-22.0-osx64.tar.gz",
+    "2cca5f99007d060aca9d8c7cbd035dfe2f040dd8200b210ce32cdf858479f70d": "bitcoin-22.0-powerpc64-linux-gnu.tar.gz",
+    "91b1e012975c5a363b5b5fcc81b5b7495e86ff703ec8262d4b9afcfec633c30d": "bitcoin-22.0-powerpc64le-linux-gnu.tar.gz",
```
It is not possible running the compat tests on `powerpc64`, because the other versions are missing on that arch, but I guess it is fine adding them here nonetheless.

MarcoFalke replied:
Maybe add a commit to remove i686 for that reason?

```diff
commit 6fd3fd1ba76831d83a43e9322e78afca70aabbeb (HEAD)
Author: MarcoFalke <falke.marco@gmail.com>
Date:   Wed Sep 22 13:35:01 2021 +0200

    test: Remove i686 from test/get_previous_releases.py
    
    It is not possible to run the compatibility tests on i686 because the
    releases v20+ are missing for that arch. It would be possible to
    self-compile those releases, but then one can also self-compile
    0.15-0.19.

diff --git a/test/get_previous_releases.py b/test/get_previous_releases.py
index 62fcad04b3..27f7044094 100755
--- a/test/get_previous_releases.py
+++ b/test/get_previous_releases.py
@@ -23,32 +23,27 @@ import hashlib
 SHA256_SUMS = {
     "d40f18b4e43c6e6370ef7db9131f584fbb137276ec2e3dba67a4b267f81cb644": "bitcoin-0.15.2-aarch64-linux-gnu.tar.gz",
     "54fb877a148a6ad189a1e1ab1ff8b11181e58ff2aaf430da55b3fd46ae549a6b": "bitcoin-0.15.2-arm-linux-gnueabihf.tar.gz",
-    "2b843506c3f1af0eeca5854a920264f9a829f02d0d50328005950ddcbe88874d": "bitcoin-0.15.2-i686-pc-linux-gnu.tar.gz",
     "87e9340ff3d382d543b2b69112376077f0c8b4f7450d372e83b68f5a1e22b2df": "bitcoin-0.15.2-osx64.tar.gz",
     "566be44190fd76daa01f13d428939dadfb8e3daacefc8fa17f433cad28f73bd5": "bitcoin-0.15.2-x86_64-linux-gnu.tar.gz",
     #
     "0768c6c15caffbaca6524824c9563b42c24f70633c681c2744649158aa3fd484": "bitcoin-0.16.3-aarch64-linux-gnu.tar.gz",
     "fb2818069854a6ad20ea03b28b55dbd35d8b1f7d453e90b83eace5d0098a2a87": "bitcoin-0.16.3-arm-linux-gnueabihf.tar.gz",
-    "75a537844313b0a84bdb61ffcdc5c4ce19a738f7ddf71007cd2edf664efd7c37": "bitcoin-0.16.3-i686-pc-linux-gnu.tar.gz",
     "78c3bff3b619a19aed575961ea43cc9e142959218835cf51aede7f0b764fc25d": "bitcoin-0.16.3-osx64.tar.gz",
     "5d422a9d544742bc0df12427383f9c2517433ce7b58cf672b9a9b17c2ef51e4f": "bitcoin-0.16.3-x86_64-linux-gnu.tar.gz",
     #
     "5a6b35d1a348a402f2d2d6ab5aed653a1a1f13bc63aaaf51605e3501b0733b7a": "bitcoin-0.17.2-aarch64-linux-gnu.tar.gz",
     "d1913a5d19c8e8da4a67d1bd5205d03c8614dfd2e02bba2fe3087476643a729e": "bitcoin-0.17.2-arm-linux-gnueabihf.tar.gz",
-    "d295fc93f39bbf0fd937b730a93184899a2eb6c3a6d53f3d857cbe77ef89b98c": "bitcoin-0.17.2-i686-pc-linux-gnu.tar.gz",
     "a783ba20706dbfd5b47fbedf42165fce70fbbc7d78003305d964f6b3da14887f": "bitcoin-0.17.2-osx64.tar.gz",
     "943f9362b9f11130177839116f48f809d83478b4c28591d486ee9a7e35179da6": "bitcoin-0.17.2-x86_64-linux-gnu.tar.gz",
     #
     "88f343af72803b851c7da13874cc5525026b0b55e63e1b5e1298390c4688adc6": "bitcoin-0.18.1-aarch64-linux-gnu.tar.gz",
     "cc7d483e4b20c5dabd4dcaf304965214cf4934bcc029ca99cbc9af00d3771a1f": "bitcoin-0.18.1-arm-linux-gnueabihf.tar.gz",
-    "989e847b3e95fc9fedc0b109cae1b4fa43348f2f712e187a118461876af9bd16": "bitcoin-0.18.1-i686-pc-linux-gnu.tar.gz",
     "b7bbcee7a7540f711b171d6981f939ca8482005fde22689bc016596d80548bb1": "bitcoin-0.18.1-osx64.tar.gz",
     "425ee5ec631ae8da71ebc1c3f5c0269c627cf459379b9b030f047107a28e3ef8": "bitcoin-0.18.1-riscv64-linux-gnu.tar.gz",
     "600d1db5e751fa85903e935a01a74f5cc57e1e7473c15fd3e17ed21e202cfe5a": "bitcoin-0.18.1-x86_64-linux-gnu.tar.gz",
     #
     "3a80431717842672df682bdb619e66523b59541483297772a7969413be3502ff": "bitcoin-0.19.1-aarch64-linux-gnu.tar.gz",
     "657f28213823d240dd3324d14829702f9ad6f0710f8bdd1c379cb3c447197f48": "bitcoin-0.19.1-arm-linux-gnueabihf.tar.gz",
-    "10d1e53208aa7603022f4acc084a046299ab4ccf25fe01e81b3fb6f856772589": "bitcoin-0.19.1-i686-pc-linux-gnu.tar.gz",
     "1ae1b87de26487075cd2fd22e0d4ead87d969bd55c44f2f1d873ecdc6147ebb3": "bitcoin-0.19.1-osx64.tar.gz",
     "aa7a9563b48aa79252c8e7b6a41c07a5441bd9f14c5e4562cc72720ea6cb0ee5": "bitcoin-0.19.1-riscv64-linux-gnu.tar.gz",
     "5fcac9416e486d4960e1a946145566350ca670f9aaba99de6542080851122e4c": "bitcoin-0.19.1-x86_64-linux-gnu.tar.gz",

Sjors replied:
When I try to `git am` this patch file it complains "Patch format detection failed.", odd.

MarcoFalke replied:
Ah, wrong format. See
[0001-test-Remove-i686-from-test-get_previous_releases.py.patch.txt](https://github.com/bitcoin/bitcoin/files/7210405/0001-test-Remove-i686-from-test-get_previous_releases.py.patch.txt)

Sjors replied:
Got it. Pushed.

Commit:
test: Remove i686 from test/get_previous_releases.py

It is not possible to run the compatibility tests on i686 because the
releases v20+ are missing for that arch. It would be possible to
self-compile those releases, but then one can also self-compile
0.15-0.19.

katesalazar commented:
Came by only to add a friendly reminder that
once this is merged, it could be interesting
to replace obsolete minor versions in a subsequent PR.

You could consider back porting as well.

I think this change could be of help for newcomers.
Thank you very much!

Sjors commented:
@katesalazar v0.21.0 is very different from v0.21.1 because the latter has taproot enabled. Not upgrading that one is intentional. For the other releases I don't think it matters too much which patch version we use here. The purpose of these tests is to find backwards compatibility issues. Ideally we would run every single patch version for that, but that might make the tests too slow.

achow101 commented on test/functional/feature_backwards_compatibility.py:
```diff
@@ -56,6 +57,7 @@ def setup_nodes(self):
         self.add_nodes(self.num_nodes, extra_args=self.extra_args, versions=[
             None,
             None,
+            22000000,
```
In a617b9d3b7ffdd7f36e8574602b182edf9d70100 "test: previous releases: add v22.0"

Previously these specifiers matched the client version number, but this does not. 22.0's client version is 220000, but this causes the test framework to search for a directory named 0.22.0 rather than 22.0.

However the client version is used in a few places in the test framework to enable features, so it would be nice if this specifier would match the actual client version. Perhaps this could use a string or tuple that represents the dotted version with a conversion within the test framework to the client version?

Sjors replied:
I think I discovered half-way this commit that 22.0 is not expressed as 22000000, but as 220000. I'll look into that.

Sjors commented:
Rebased and changed the client version in the test from `22000000` to `220000`.

katesalazar commented:
I wonder if 9e03b891af is enough self explanatory
about the bugs it fixes.

It wouldn't hurt me adding a thorough detailed text to it.

Sjors commented:
@katesalazar that commit ensures that we we add a new version to the list, everything keeps working. It also adds some missing node_master.unloadwallet statements. I wrote it more than a year ago, so that's all I remember :-)

Sjors committed:
test: bump sandbox argument minimum version

The -sandbox argument is not present in the v22.0 release. Changing the minimum version to 229900 ensures it's used when testing the master branch.

If the argument is backported, the minimum version can be adjusted to e.g. 220100.

Sjors commented:
The test framework sets `-sandbox=log-and-abort` for version 22.0 and up, but that's not part of the release binary. I added a commit to fix that.

Sjors commented:
Rebased after #23515 and merge conflict.

ryanofsky commented on test/functional/feature_backwards_compatibility.py:
```diff
@@ -156,173 +150,72 @@ def run_test(self):
-                os.path.join(node_master_wallets_dir, wallet),
-                os.path.join(node_v19_wallets_dir, wallet)
-            )
+        for node in self.nodes[2:]:
```
In commit "test: v0.20.1 backwards compatibility" (94d3989352aba63c86c901269438bfd566dd8632)

The deduplication here is great but the `nodes[2:]` references are more obscure and maybe fragile. Could this use `for node in legacy_nodes`? And

legacy_nodes = [node_v19, node_v18, node_v17]
or
legacy_nodes = nodes[2:]
or
legacy_nodes = [node for node in nodes if node.version <= 190000]

?

Sjors replied:
That's a good idea...

ryanofsky commented on test/functional/feature_backwards_compatibility.py:
```diff
@@ -156,173 +150,72 @@ def run_test(self):
+                # and loadwallet is introduced in v0.17.0
+                if node.version >= 170000 and node.version < 210000:
+                    for wallet_name in ["w1", "w2", "w3"]:
+                        assert_raises_rpc_error(-4, "Wallet file verification failed: wallet.dat corrupt, salvage failed", node.loadwallet, wallet_name)
```
In commit "test: v0.20.1 backwards compatibility" (94d3989352aba63c86c901269438bfd566dd8632)

Looking at all the test changes here it seems like no test coverage is lost, but it would be nice if commit message could say whether this is the case.

IMO, it would also be nicer if this commit were split into a refactoring commit that did not change coverage and just deduped code, and a smaller followup commit adding the new v0.20 version. But just mentioning however this commit affects coverage would be good if you don't want to change the commit contents.

Sjors replied:
I moved the refactoring stuff into the misc changes commit. Adding a note that perhaps coverage changed in the process of this cleanup.

ryanofsky commented on test/functional/test_framework/test_framework.py:
```diff
@@ -447,11 +447,13 @@ def add_nodes(self, num_nodes: int, extra_args=None, *, rpchost=None, binary=Non
         def get_bin_from_version(version, bin_name, bin_default):
             if not version:
                 return bin_default
+            if version > 219999:
+                version *= 100
```
In commit "test: previous releases: add v22.0" (62b22de69dabfb86eaabb6a685e3989ce525d0e3)

This seems a little strange. Could you add a code comment here explaining this? It seems like this code could not have a special 219999 case and just multiply by 100 for all versions and strip .0.0 for all versions.

Sjors replied:
I'll add a note. When we changed the numbering in version v22.0, the client version wasn't bumped to 22000000, but to 220000. Without this change it would expect the binary to live in v0.22.0. Conversely if we multiplied all versions by 100 then it would look for v21.0 for the 0.21.0 release.

ryanofsky reviewed (approved):
Code review ACK 62b22de69dabfb86eaabb6a685e3989ce525d0e3. This all looks very good, and the deduping is nice. Left suggestions below but they aren't important and are just questions and requests to explain some things better.

katesalazar commented on test/functional/feature_backwards_compatibility.py:
```diff
@@ -72,7 +72,7 @@ def run_test(self):
         assert_equal(res['blocks'], COINBASE_MATURITY + 1)
 
-        node_master = self.nodes[self.num_nodes - 5]
+        node_master = self.nodes[1]
```
What I [was trying to tell][0], clearly expressed.

[0]: https://github.com/bitcoin/bitcoin/pull/19013#issuecomment-968349233

Sjors replied:
I added a bit more clarification. In particular I point out that coverage may have changed, which I think is worth it for the increased readability.

Sjors committed:
test: v0.20.1 backwards compatibility

The file checksums were added in an earlier commit. Since the DMG
file is never downloaded, we drop that checksum.

Sjors commented on test/get_previous_releases.py:
```diff
@@ -51,7 +51,6 @@
     "60c93e3462c303eb080be7cf623f1a7684b37fd47a018ad3848bc23e13c84e1c": "bitcoin-0.20.1-aarch64-linux-gnu.tar.gz",
     "55b577e0fb306fb429d4be6c9316607753e8543e5946b542d75d876a2f08654c": "bitcoin-0.20.1-arm-linux-gnueabihf.tar.gz",
     "b9024dde373ea7dad707363e07ec7e265383204127539ae0c234bff3a61da0d1": "bitcoin-0.20.1-osx64.tar.gz",
-    "c378d4e21109f09e8829f3591e015c66632dff2925a60b64d259be05a334c30b": "bitcoin-0.20.1-osx.dmg",
```
Yes, the DMG was added in an earlier PR, but shouldn't have. I'll add a note.

Sjors commented:
Rebased, moved refactoring stuff out of the v0.20 commit into 408e9b4. I improved that commit further by using variables in more places instead of indexes.

ryanofsky reviewed (approved):
Code review ACK 7124899057a351a3491b288d96426f91ef1fd040. This PR looks great and now is simpler to review with code changes consolidated in one commit, and new releases added in other commits. Other than that, only other changes since last review are replacing hardcoded node indexes with variables, and adding a code and commit comments

Sjors commented:
Thanks @ryanofsky. I fixed the indentation and added the requested comment.

Sjors commented:
Rebased and included #19332.

ryanofsky reviewed (approved):
Code review ACK 24cec4b5c02e12cf0b6b56ba5055b8f5758627a5. Only change since last review is rebasing and adding comment and whitelist args.

I'm surprised this PR is still open. It seems useful and not a particularly risky change to merge.

Sjors commented:
🎉
